server/etcdserver/txn/metrics.go

```diff
@@ -39,13 +39,29 @@ var (
 		Buckets: prometheus.ExponentialBuckets(0.0001, 2, 20),
 	},
 		[]string{"version", "op", "success"})
+	rangeSec = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "etcd",
+		Subsystem: "server",
+		Name:      "range_duration_seconds",
+		Help:      "The latency distributions of txn.Range",
+
+		// lowest bucket start of upper bound 0.0001 sec (0.1 ms) with factor 2
+		// highest bucket start of 0.0001 sec * 2^19 == 52.4288 sec
+		Buckets: prometheus.ExponentialBuckets(0.0001, 2, 20),
+	},
+		[]string{"success"})
 )
 
 func ApplySecObserve(version, op string, success bool, latency time.Duration) {
 	applySec.WithLabelValues(version, op, strconv.FormatBool(success)).Observe(float64(latency.Microseconds()) / 1000000.0)
 }
 
+func RangeSecObserve(op string, success bool, latency time.Duration) {
```

Comment:
Can we remove `op` if we aren't using it in `rangeSec`?

```suggestion
func RangeSecObserve(success bool, latency time.Duration) {
```

Reply:
:eyes: :thinking: This should have been caught by the `unparam` linter.

Reply:
Ah, I think it falls under an `unparam` exception:

> To minimise false positives, it ignores certain cases such as:
> * Exported functions (by default, see `-exported`)
...

If I ran the linter with `-exported,` it did find the issue. However, it will open a can of worms. We'll need to go through the deprecation process with changes to many exported functions. Refer to the following list of all the offenders:

```
'lint' started at Sun May 12 04:43:37 PM PDT 2024
% (cd api && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
% (cd pkg && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
FAIL: (code:1):
  % (cd pkg && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
flags/ignored.go:32:27: `(*IgnoredFlag).Set` - `s` is unused (unparam)
func (f *IgnoredFlag) Set(s string) error {
                          ^
flags/strings.go:29:39: (*StringsValue).Set - result 0 (error) is always nil (unparam)
func (ss *StringsValue) Set(s string) error {
                                      ^
flags/unique_strings.go:33:45: (*UniqueStringsValue).Set - result 0 (error) is always nil (unparam)
func (us *UniqueStringsValue) Set(s string) error {
                                            ^
cobrautl/help.go:146:64: UsageFunc - result 0 (error) is always nil (unparam)
func UsageFunc(cmd *cobra.Command, version, APIVersion string) error {
                                                               ^
grpctesting/stub_server.go:56:56: `(*StubServer).Start` - `dopts` is unused (unparam)
func (ss *StubServer) Start(sopts []grpc.ServerOption, dopts ...grpc.DialOption) error {
                                                       ^
grpctesting/stub_server.go:114:25: `NewDummyStubServer` - `body` is unused (unparam)
func NewDummyStubServer(body []byte) *StubServer {
                        ^
FAIL: 'run golangci-lint run --config /home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml ./...' checking failed (!=0 return code)
There was a Failure in module pkg, keep going...
% (cd client/pkg && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
FAIL: (code:1):
  % (cd client/pkg && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
testutil/testutil.go:102:22: `ExitInShortMode` - `reason` is unused (unparam)
func ExitInShortMode(reason string) {
                     ^
FAIL: 'run golangci-lint run --config /home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml ./...' checking failed (!=0 return code)
There was a Failure in module client/pkg, keep going...
% (cd client/internal/v2 && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
% (cd client/v3 && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
FAIL: (code:1):
  % (cd client/v3 && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
retry.go:191:29: `RetryMaintenanceClient` - `c` is unused (unparam)
func RetryMaintenanceClient(c *Client, conn *grpc.ClientConn) pb.MaintenanceClient {
                            ^
naming/resolver/resolver.go:67:62: NewBuilder - result 1 (error) is always nil (unparam)
func NewBuilder(client *clientv3.Client) (gresolver.Builder, error) {
                                                             ^
experimental/recipes/watch.go:25:91: WaitEvents - result 0 (*go.etcd.io/etcd/client/v3.Event) is never used (unparam)
func WaitEvents(c *clientv3.Client, key string, rev int64, evs []mvccpb.Event_EventType) (*clientv3.Event, error) {
                                                                                          ^
FAIL: 'run golangci-lint run --config /home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml ./...' checking failed (!=0 return code)
There was a Failure in module client/v3, keep going...
% (cd server && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
FAIL: (code:1):
  % (cd server && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
etcdserver/api/membership/cluster.go:684:46: `(*RaftCluster).IsReadyToPromoteMember` - `id` is unused (unparam)
func (c *RaftCluster) IsReadyToPromoteMember(id uint64) bool {
                                             ^
auth/store.go:319:30: (*authStore).Close - result 0 (error) is always nil (unparam)
func (as *authStore) Close() error {
                             ^
auth/store.go:329:66: `(*authStore).Authenticate` - `password` is unused (unparam)
func (as *authStore) Authenticate(ctx context.Context, username, password string) (*pb.AuthenticateResponse, error) {
                                                                 ^
auth/store.go:358:64: (*authStore).CheckPassword - result 0 (uint64) is never used (unparam)
func (as *authStore) CheckPassword(username, password string) (uint64, error) {
                                                               ^
auth/store.go:420:57: (*authStore).UserAdd - result 0 (*go.etcd.io/etcd/api/v3/etcdserverpb.AuthUserAddResponse) is never used (unparam)
func (as *authStore) UserAdd(r *pb.AuthUserAddRequest) (*pb.AuthUserAddResponse, error) {
                                                        ^
auth/store.go:465:63: (*authStore).UserDelete - result 0 (*go.etcd.io/etcd/api/v3/etcdserverpb.AuthUserDeleteResponse) is never used (unparam)
func (as *authStore) UserDelete(r *pb.AuthUserDeleteRequest) (*pb.AuthUserDeleteResponse, error) {
                                                              ^
auth/store.go:494:79: (*authStore).UserChangePassword - result 0 (*go.etcd.io/etcd/api/v3/etcdserverpb.AuthUserChangePasswordResponse) is never used (unparam)
func (as *authStore) UserChangePassword(r *pb.AuthUserChangePasswordRequest) (*pb.AuthUserChangePasswordResponse, error) {
                                                                              ^
auth/store.go:536:69: (*authStore).UserGrantRole - result 0 (*go.etcd.io/etcd/api/v3/etcdserverpb.AuthUserGrantRoleResponse) is never used (unparam)
func (as *authStore) UserGrantRole(r *pb.AuthUserGrantRoleRequest) (*pb.AuthUserGrantRoleResponse, error) {
                                                                    ^
auth/store.go:593:31: `(*authStore).UserList` - `r` is unused (unparam)
func (as *authStore) UserList(r *pb.AuthUserListRequest) (*pb.AuthUserListResponse, error) {
                              ^
auth/store.go:668:31: `(*authStore).RoleList` - `r` is unused (unparam)
func (as *authStore) RoleList(r *pb.AuthRoleListRequest) (*pb.AuthRoleListResponse, error) {
                              ^
auth/store.go:762:57: (*authStore).RoleAdd - result 0 (*go.etcd.io/etcd/api/v3/etcdserverpb.AuthRoleAddResponse) is never used (unparam)
func (as *authStore) RoleAdd(r *pb.AuthRoleAddRequest) (*pb.AuthRoleAddResponse, error) {
                                                        ^
auth/store.go:806:81: (*authStore).RoleGrantPermission - result 0 (*go.etcd.io/etcd/api/v3/etcdserverpb.AuthRoleGrantPermissionResponse) is never used (unparam)
func (as *authStore) RoleGrantPermission(r *pb.AuthRoleGrantPermissionRequest) (*pb.AuthRoleGrantPermissionResponse, error) {
                                                                                ^
proxy/grpcproxy/adapter/chan_stream.go:92:41: (*chanClientStream).CloseSend - result 0 (error) is always nil (unparam)
func (cs *chanClientStream) CloseSend() error {
                                        ^
lease/lessor.go:829:41: `(*FakeLessor).Grant` - `ttl` is unused (unparam)
func (fl *FakeLessor) Grant(id LeaseID, ttl int64) (*Lease, error) {
                                        ^
etcdserver/txn/metrics.go:59:22: `RangeSecObserve` - `op` is unused (unparam)
func RangeSecObserve(op string, success bool, latency time.Duration) {
                     ^
etcdserver/txn/txn.go:252:135: Txn - result 0 (*go.etcd.io/etcd/api/v3/etcdserverpb.TxnResponse) is never used (unparam)
func Txn(ctx context.Context, lg *zap.Logger, rt *pb.TxnRequest, txnModeWriteWithSharedBuffer bool, kv mvcc.KV, lessor lease.Lessor) (*pb.TxnResponse, *traceutil.Trace, error) {
                                                                                                                                      ^
etcdserver/server_test.go:1338:32: `(*nodeRecorder).Compact` - `index` is unused (unparam)
func (n *nodeRecorder) Compact(index uint64, nodes []uint64, d []byte) {
                               ^
etcdserver/v3_server.go:451:105: (*EtcdServer).LeaseLeases - result 1 (error) is always nil (unparam)
func (s *EtcdServer) LeaseLeases(_ context.Context, _ *pb.LeaseLeasesRequest) (*pb.LeaseLeasesResponse, error) {
                                                                                                        ^
etcdserver/api/v3rpc/maintenance.go:108:41: `(*maintenanceServer).Defragment` - `ctx` is unused (unparam)
func (ms *maintenanceServer) Defragment(ctx context.Context, sr *pb.DefragmentRequest) (*pb.DefragmentResponse, error) {
                                        ^
etcdserver/api/v3rpc/maintenance.go:124:39: `(*maintenanceServer).Snapshot` - `sr` is unused (unparam)
func (ms *maintenanceServer) Snapshot(sr *pb.SnapshotRequest, srv pb.Maintenance_SnapshotServer) error {
                                      ^
etcdserver/api/v3rpc/maintenance.go:212:35: `(*maintenanceServer).Hash` - `ctx` is unused (unparam)
func (ms *maintenanceServer) Hash(ctx context.Context, r *pb.HashRequest) (*pb.HashResponse, error) {
                                  ^
etcdserver/api/v3rpc/maintenance.go:222:37: `(*maintenanceServer).HashKV` - `ctx` is unused (unparam)
func (ms *maintenanceServer) HashKV(ctx context.Context, r *pb.HashKVRequest) (*pb.HashKVResponse, error) {
                                    ^
etcdserver/api/v3rpc/maintenance.go:250:37: `(*maintenanceServer).Status` - `ctx` is unused (unparam)
func (ms *maintenanceServer) Status(ctx context.Context, ar *pb.StatusRequest) (*pb.StatusResponse, error) {
                                    ^
FAIL: 'run golangci-lint run --config /home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml ./...' checking failed (!=0 return code)
There was a Failure in module server, keep going...
% (cd etcdutl && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
% (cd etcdctl && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
% (cd tests && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
FAIL: (code:1):
  % (cd tests && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
framework/e2e/etcdctl.go:346:88: (*EtcdctlV3).Compact - result 0 (*go.etcd.io/etcd/client/v3.CompactResponse) is always nil (unparam)
func (ctl *EtcdctlV3) Compact(ctx context.Context, rev int64, o config.CompactOption) (*clientv3.CompactResponse, error) {
                                                                                       ^
framework/e2e/util.go:61:122: SpawnWithExpectLines - result 0 ([]string) is never used (unparam)
func SpawnWithExpectLines(ctx context.Context, args []string, envVars map[string]string, xs ...expect.ExpectedResponse) ([]string, error) {
                                                                                                                         ^
framework/integration/cluster.go:1130:30: `(*Member).WaitStarted` - `t` is unused (unparam)
func (m *Member) WaitStarted(t testutil.TB) {
                             ^
integration/v2store/store_test.go:818:33: `TestStoreWatchSlowConsumer` - `t` is unused (unparam)
func TestStoreWatchSlowConsumer(t *testing.T) {
                                ^
FAIL: 'run golangci-lint run --config /home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml ./...' checking failed (!=0 return code)
There was a Failure in module tests, keep going...
% (cd tools/rw-heatmaps && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
% (cd tools/testgrid-analysis && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
FAIL: (code:1):
  % (cd tools/testgrid-analysis && 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...')
cmd/github.go:33:20: `getOpenIssues` - `tab` is unused (unparam)
func getOpenIssues(tab string, labels []string) []*github.Issue {
                   ^
FAIL: 'run golangci-lint run --config /home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml ./...' checking failed (!=0 return code)
There was a Failure in module tools/testgrid-analysis, keep going...
% 'golangci-lint' 'run' '--config' '/home/ivan/Code/Personal/etcd/etcd/tools/.golangci.yaml' './...'
FAIL: 'lint' FAILED at Sun May 12 04:44:22 PM PDT 2024
make: *** [Makefile:82: verify-lint] Error 255
```